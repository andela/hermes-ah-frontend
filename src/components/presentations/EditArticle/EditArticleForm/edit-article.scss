@import '../../../../assets/stylesheets/index.scss';

#whole-form {
    margin-top: 90px;
}
.input-heading{
    margin-left: 8.3%;
}
.input-heading2{
    margin-left: 30.7%;
}
.image-space{
    color: blue;
    background-color: black;
}

.input-row{
    .input-holders {
    margin-left: 8.5%;
  }
  .ui.input {
    width: 45%;
}
.ui.input>input {
    width: 45%;
    border-radius: 0;
  }
  textarea {
    font-family: sans-serif;
    font-size: 100%;
    line-height: 1.15;
    padding: .67857143em 1em;
    height: 124px;
    margin: 0;
    width: 83%;
    border: solid 1px rgba(34,36,38,.15);
    }
}

.editor-group {
    .rdw-editor-wrapper {
        box-sizing: content-box;
        width: 83%;
        height: 500px;
        margin: 0 auto;
        border: 1px rgb(206, 206, 206) solid;
    }
    .rdw-image-modal {
        position: absolute;
        top: 35px;
        left: -200px;
        display: flex;
        flex-direction: column;
        width: 235px;
        border: 1px solid rgb(241, 241, 241);
        padding: 15px;
        border-radius: 2px;
        z-index: 100;
        background: rgb(255, 255, 255);
        box-shadow: 3px 3px 5px rgb(191, 189, 189);
    }
    
    .rdw-editor-main {
        height: 370px;
        overflow: auto;
        padding: .67857143em 1em;
    }
    .DraftEditor-editorContainer{
        height: initial;
    }
}

.dropdown-group{
    .keyword-instruction {
        margin-left: 8.5%;
        color: rgb(124, 122, 122);
    }
    
    .ui.fluid.dropdown {
        display: block;
        width: 50%;
        margin-left: 8.5%;
        min-width: 0;
    }
}

#header-image {
    border: solid 1px rgb(0, 0, 255);
    height: 50px;
}
.image-group{
    .image-upload {
        margin-left: 8.5%;
        padding: 1% 0 0 2%;
        width: 35%;
        height: 130px;
        border: 1px solid rgb(231, 230, 230);
      }
    
       .image-upload p {
        margin: 0;
      }
    
       .image-upload-box {
        background-color: rgb(235, 235, 235);
        padding: 0 auto;
        width: 90%;
        height: 50px;
        border: 1px solid rgb(192, 191, 191);
    }
    
     .image-upload-box input[type="file"] {
        margin: 14px 16% 0;
    }
    .input-file-heading {
        margin-left: 30.3%;
    }
}

.text-center {
    .ui.dropdown:not(.button) > .default.text {
        color: rgba(191, 191, 191, 0.87);
        font-size: 100%;
    }
    
     .ui.button{
        margin-left: 1%;
        background-color: $primary-color;
        color: rgb(255, 255, 255);
     } 
     
     .ui.button:nth-child(1){
        margin-left: 9%;
     }
     .cancel-button {
        background-color:$primary-color; 
        color:rgba(255, 255, 255, 0.15); 
     }
     
    .ui.basic.blue.button {
         color: $primary-color; 
         margin: 1%;
     }
     .ui.button:hover {
        background-color:rgba(255, 255, 255, 0.15); 
        color:$primary-color; 
        border: 0.5px solid $primary-color; 
    }
}

 .style-button, #style-button{
     padding: 1%;
     border: 1px solid rgb(221, 221, 221);
 }
 
  .style-button:hover{
     border: solid 1px $primary-color;
     background-color: rgb(231, 230, 230);
 }

@media screen and (max-width: 480px){
.text-center {
    font-size: 19px;
    }
}

@media screen and (max-width: 854px){
    .image-group{
        .image-upload {
            width: 50%;
          }
       }
    }

@media screen and (max-width: 577px){
    .image-group{
        .image-upload {
            width: 65%;
          }
    }
    .editor-group {
        .rdw-editor-main {
            height: 350px;
        }
    }
        .text-center {
        .ui.button{
            margin-left: 2%;
         }
         .ui.button:nth-child(1){
            margin-left: 18%;
         }
         .ui.button:nth-child(3){
            margin-left: 40%;
            margin-top: 20px;
         }
      }
    }

        @media screen and (max-width: 442px){
            .image-upload {
                width: 75%;
              }
              .text-center{
              .ui.button{
                margin-left: 2%;
             }
             .ui.button:nth-child(1){
                margin-left: 18%;
             }
             .ui.button:nth-child(3){
                margin-left: 40%;
                margin-top: 20px;
             }
            }
        }

            @media screen and (max-width: 392px){
                .editor-group{
                    .rdw-editor-main {
                        height: 300px;
                    }
                }
                .image-upload {
                    width: 100%;
                  }
                  .text-center{
                  .ui.button{
                    margin-left: 2%;
                 }
                 .ui.button:nth-child(1){
                    margin-left: 40%;
                 }
                 .ui.button:nth-child(2){
                    margin-top: 20px;
                    margin-left: 38.6%;
                 }
                 .ui.button:nth-child(3){
                    margin-left: 46%;
                    margin-top: 20px;
                 }
                }
            }
                @media screen and (max-width: 213px){
                    #form-whole {
                        margin: 0 auto;
                    }
                    .input-heading{
                        text-align: center;
                        margin-left: 0;
                    }
                    .image-upload {
                        width: 100%;
                      }

                      .editor-group{
                        .rdw-editor-main {
                            height: 300px;
                        }
                      }
                      .input-row{  
                          .ui.input {
                        margin-left: 25%;
                    }
                }

                    .text-center{
                    .ui.fluid.dropdown {
                        margin-left: 25%;
                    }
                      .ui.button{
                        margin-left: 2%;
                     }
                     .ui.button:nth-child(1){
                        margin-left: 40%;
                     }
                     .ui.button:nth-child(2){
                        margin-top: 20px;
                        margin-left: 38.6%;
                     }
                     .ui.button:nth-child(3){
                        margin-left: 46%;
                        margin-top: 20px;
                     }
                }
            }
