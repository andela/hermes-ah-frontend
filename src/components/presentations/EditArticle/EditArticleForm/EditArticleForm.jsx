import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Editor } from 'react-draft-wysiwyg';
import 'react-draft-wysiwyg/dist/react-draft-wysiwyg.css';
import { Input, TextArea, Dropdown, Button } from 'semantic-ui-react';
import categoryoptions from '../../NewArticle/NewArticleForm/category';
import './edit-article.scss';

const EditArticleForm = ({
  saveCategory,
  saveToCloudinary,
  editorState,
  defaultEditorState,
  onEditorStateChange,
  saveOrPublish,
  onChange,
  headerImage,
  currentValues,
  handleChange,
  handleAddition,
  keywords,
  titleRaw,
  abstractRaw,
  categoryRaw,
}) => (
  <div>
    <main className="main-content">
      <section className="section">
        <div className="container">
          <div className="row">
            <div className="col-12 col-lg-12">
              <form className="p-30 bg-gray rounded" id="whole-form">
                <div className="input-row">
                  <div>
                    <h3 className="input-heading">Title</h3>
                    <Input
                      className="input-holders"
                      name="title"
                      defaultValue={titleRaw}
                      onChange={onChange}
                    />
                  </div>
                  <div>
                    <br />
                    <h3 className="input-heading">Abstract</h3>
                    <TextArea
                      className="input-holders"
                      value={abstractRaw}
                      onChange={onChange}
                      name="abstract"
                    />
                  </div>
                </div>
                <br />
                <div className="editor-group">
                  <h3 className="input-heading">Body</h3>
                  <Editor
                    onEditorStateChange={onEditorStateChange}
                    editorState={editorState}
                    defaultEditorState={defaultEditorState}
                  />
                </div>
                <br />
                <div className="dropdown-group">
                  <div>
                    <h3 className="input-heading">Keywords</h3>
                    <Dropdown
                      options={keywords}
                      placeholder="Type in your keywords"
                      search
                      selection
                      fluid
                      multiple
                      allowAdditions
                      onAddItem={handleAddition}
                      onChange={handleChange}
                      value={currentValues}
                    />
                    <p className="keyword-instruction">
                      Enter a comma-separated list. For example: Anatomy,
                      Bioinformatics
                    </p>
                  </div>
                  <br />
                  <div>
                    <h3 className="input-heading">Category</h3>
                    <Dropdown
                      placeholder="Select category"
                      fluid
                      search
                      selection
                      options={categoryoptions}
                      onChange={saveCategory}
                      className="input-holders"
                      defaultValue={categoryRaw}
                    />
                  </div>
                </div>
                <br />
                <h3 className="input-heading">Article Header Image</h3>
                <div className="image-group">
                  <div className="image-upload">
                    <div className="image-upload-box">
                      <input
                        type="file"
                        name="image"
                        className="form-control ft"
                        onChange={saveToCloudinary}
                        accept=".png, .jpg, .jpeg"
                      />
                    </div>
                    <p>One file only</p>
                    <p>2 MB limit</p>
                    <p>Allowed types: png, jpg, jpeg</p>
                  </div>
                  <div className="headerImage">
                    <img alt="" src={headerImage} />
                  </div>
                </div>
                <br />
                <br />
                <br />
                <div className="text-center">
                  <Button
                    className="save-button"
                    type="submit"
                    onClick={e => saveOrPublish(e, true)}
                  >
                    Save as Draft
                  </Button>
                  <Button
                    className="save-button"
                    type="submit"
                    onClick={e => saveOrPublish(e, false)}
                  >
                    Save & Publish
                  </Button>
                  <Link to="/profile/">
                    <Button className="save-button" basic color="blue">
                      Cancel
                    </Button>
                  </Link>
                </div>
                <br />
                <br />
              </form>
            </div>
          </div>
        </div>
      </section>
    </main>
  </div>
);
EditArticleForm.propTypes = {
  onChange: PropTypes.func.isRequired,
  saveOrPublish: PropTypes.func.isRequired,
  onEditorStateChange: PropTypes.func.isRequired,
  editorState: PropTypes.shape().isRequired,
  defaultEditorState: PropTypes.shape().isRequired,
  saveToCloudinary: PropTypes.func.isRequired,
  saveCategory: PropTypes.func.isRequired,
  headerImage: PropTypes.string.isRequired,
  titleRaw: PropTypes.string.isRequired,
  categoryRaw: PropTypes.string.isRequired,
  keywords: PropTypes.arrayOf(PropTypes.string).isRequired,
  currentValues: PropTypes.arrayOf(PropTypes.string).isRequired,
  handleChange: PropTypes.func.isRequired,
  handleAddition: PropTypes.func.isRequired,
  abstractRaw: PropTypes.string.isRequired,
};

export default EditArticleForm;
