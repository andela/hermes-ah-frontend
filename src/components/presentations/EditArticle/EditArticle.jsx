import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { EditorState, convertToRaw, ContentState } from 'draft-js';
import { toast } from 'react-toastify';
import draftToHtml from 'draftjs-to-html';
import htmlToDraft from 'html-to-draftjs';
import uploadToCloudnary from '../../../utils/uploadToCloudnary';
import EditArticleForm from './EditArticleForm/EditArticleForm';
import validateImage from '../../../utils/validateImage';
import NavBar from '../../shared/NavBar/NavBar';

class EditArticle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      category: '',
      options: [],
      imageUrl: '',
      editorState: EditorState.createEmpty(),
      defaultEditorState: '',
      body: '',
      updatedAbstract: false,
      title: '',
      abstract: '',
      currentValues: [''],
    };
    this.onChange = this.onChange.bind(this);
    this.onEditorStateChange = this.onEditorStateChange.bind(this);
  }

  componentDidMount = () => {
    const { getSingleArticle, match } = this.props;
    const { articleId } = match.params;
    return getSingleArticle(articleId);
  };

  static getDerivedStateFromProps(props, state) {
    if (
      props.abstract &&
      props.singleArticle.article.Keywords &&
      state.updatedAbstract === false
    ) {
      const blocksFromHtml = htmlToDraft(props.singleArticle.article.body);
      const { contentBlocks, entityMap } = blocksFromHtml;
      const contentState = ContentState.createFromBlockArray(
        contentBlocks,
        entityMap
      );
      const { singleArticle } = props;
      const { article } = singleArticle;
      const { category } = article;

      return {
        abstract: props.abstract,
        updatedAbstract: true,
        keywords: props.singleArticle.article.Keywords,
        options: props.singleArticle.article.Keywords.map(option => {
          return {
            key: option.keyword,
            text: option.keyword,
            value: option.keyword,
          };
        }),
        category: [{ key: category, text: category, value: category }],
        currentValues: props.singleArticle.article.Keywords.map(
          option => option.keyword
        ),
        title: props.singleArticle.article.title,
        imageUrl: props.singleArticle.article.image_url,
        editorState: EditorState.createWithContent(contentState),
        body: props.singleArticle.article.body,
      };
    }
    return state;
  }

  onChange(e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  onEditorStateChange = editorState => {
    const body = draftToHtml(convertToRaw(editorState.getCurrentContent()));
    this.setState({
      body,
      editorState,
    });
  };

  handleAddition = (_e, { value }) => {
    this.setState(prevState => ({
      options: [{ text: value, value }, ...prevState.options],
    }));
  };

  handleChange = (e, { value }) => this.setState({ currentValues: value });

  saveToCloudinary = async e => {
    const form = new FormData();
    const imageData = e.target.files[0];
    const validFormat = validateImage(imageData);
    if (validFormat.valid) {
      toast.info(validFormat.message, {
        type: toast.TYPE.INFO,
        closeButton: false,
        position: toast.POSITION.TOP_CENTER,
      });
      form.append('file', imageData);
      const res = await uploadToCloudnary(form);
      this.setState({ imageUrl: res.url });
      toast.dismiss();
    } else {
      toast.error(validFormat.message);
      e.target.value = '';
    }
  };

  saveCategory = e => {
    e.preventDefault();
    const categoryData = e.target.attributes.getNamedItem('name').value;
    this.setState({ category: categoryData });
  };

  saveOrPublish = async (e, isDraft) => {
    e.preventDefault();
    const { title, abstract, imageUrl, body, keywords, category } = this.state;
    const data = {
      title,
      abstract,
      is_draft: isDraft,
      image_url: imageUrl,
      body,
      keywords: keywords.map(keyword => keyword.keyword),
      category,
    };
    const { editAnArticle, match } = this.props;
    const { articleId } = match.params;
    await editAnArticle(articleId, data);
  };

  render() {
    const {
      editorState,
      options,
      title,
      abstract,
      category,
      currentValues,
      imageUrl,
      defaultEditorState,
    } = this.state;
    const { singleArticle } = this.props;
    const { article } = singleArticle;
    return (
      <React.Fragment>
        <NavBar />
        <EditArticleForm
          saveCategory={this.saveCategory}
          titleRaw={title}
          valueRaw={article.abstract}
          bodyRaw={article.editorState}
          abstractRaw={abstract}
          categoryRaw={category}
          onEditorStateChange={this.onEditorStateChange}
          editorState={editorState}
          defaultEditorState={defaultEditorState}
          currentValues={currentValues}
          onChange={this.onChange}
          saveOrPublish={this.saveOrPublish}
          saveToCloudinary={this.saveToCloudinary}
          headerImage={imageUrl}
          handleChange={this.handleChange}
          handleAddition={this.handleAddition}
          keywords={options}
          sampleEditorContent={editorState}
        />
      </React.Fragment>
    );
  }
}

EditArticle.propTypes = {
  getSingleArticle: PropTypes.func.isRequired,
  editAnArticle: PropTypes.func.isRequired,
  match: PropTypes.shape(PropTypes.objectOf).isRequired,
  singleArticle: PropTypes.shape({}).isRequired,
};

export default EditArticle;
