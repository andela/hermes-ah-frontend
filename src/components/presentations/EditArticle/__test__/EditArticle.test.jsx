import React from 'react';
// import { MemoryRouter as Router } from 'react-router-dom';
import { shallow } from 'enzyme';
import EditArticleForm from '../EditArticleForm/EditArticleForm';
import EditArticle from '../EditArticle';

describe('editArticle component', () => {
  it('should call edit onChange prop', () => {
    const props = {
      history: {
        push: '/',
      },
      singleArticle: { article: [] },
      match: { params: { articleId: '2' } },
      getSingleArticle: jest.fn(),
    };
    const event = {
      preventDefault: jest.fn(),
      target: { abstract: 'title', value: 'The-value' },
    };
    const component = shallow(<EditArticle {...props} />);
    expect(component.instance().state.abstract).toEqual('');
    component.instance().onChange(event);
  });

  it('should call edit onChange prop', () => {
    const props = {
      history: {
        push: '/',
      },
      singleArticle: { article: [] },
      match: { params: { articleId: '2' } },
      getSingleArticle: jest.fn(),
      saveToCloudinary: jest.fn(),
      uploadToCloudnary: jest.fn(() => Promise.resolve()),
    };
    const event = {
      preventDefault: jest.fn(),
      target: {
        files: [
          {
            lastModified: 1556711757797,
            lastModifiedDate: '',
            name: 'books-bookshelves-bookstore-1184589.jpg',
            size: 1979381,
            type: 'image/jpeg',
            webkitRelativePath: '',
          },
        ],
      },
    };
    const component = shallow(<EditArticle {...props} />);

    // console.log(component.debug());
    expect(component.instance().saveToCloudinary(event));

    // expect(component.instance().state.abstract).toEqual('');
    // component.instance().onChange(event);
  });

  it('should call edit onChange prop', () => {
    const props = {
      history: {
        push: '/',
      },
      singleArticle: { article: [] },
      match: { params: { articleId: '2' } },
      getSingleArticle: jest.fn(),
      saveToCloudinary: jest.fn(),
    };
    const event = {
      preventDefault: jest.fn(),
      target: {
        files: 'gdgdg',
      },
    };
    const component = shallow(<EditArticle {...props} />);

    // console.log(component.debug());
    expect(component.instance().saveToCloudinary(event));

    // expect(component.instance().state.abstract).toEqual('');
    // component.instance().onChange(event);
  });

  it('should call handleAddition prop', () => {
    const props = {
      postArticle: jest.fn(),
      success: true,
      history: {
        push: '/',
      },
      singleArticle: { article: [] },
      match: { params: { articleId: '2' } },
      getSingleArticle: jest.fn(),
    };
    const event = {
      preventDefault: jest.fn(),
      target: { options: [{ text: 'title', value: 'title' }] },
    };
    const component = shallow(<EditArticle {...props} />);
    component.instance().handleAddition(event, { value: 'title' });
    expect(component.state().options[0].value).toBe('title');
  });

  it('should call handleChange prop', () => {
    const props = {
      postArticle: jest.fn(),
      success: true,
      history: {
        push: '/',
      },
      singleArticle: { article: [] },
      match: { params: { articleId: '2' } },
      getSingleArticle: jest.fn(),
    };
    const event = {
      preventDefault: jest.fn(),
      target: {
        keywords: [{ value: 'title' }],
      },
    };
    const component = shallow(<EditArticle {...props} />);
    component.instance().handleChange(event, { value: ['title'] });
  });

  it('should call saveOrPublish prop', () => {
    const article = {
      title: "Scientist don't die",
      abstract: 'Science is always alive',
      imageUrl:
        'https://res.cloudinary.com/dnch08bzc/image/upload/v1556821917/3d-rendering-human-brain-technology-background_34663-72.jpg',
      body: 'Lorem ipsum lorem ipsum lorem ipsum',
      keywords: ['science', 'fiction', 'vision'],
      category: 'Science',
    };

    const props = {
      postArticle: jest
        .fn()
        .mockReturnValue(Promise.resolve({ data: article })),
      success: true,
      history: {
        push: '/',
      },
      singleArticle: { article: [] },
      match: { params: { articleId: '2' } },
      getSingleArticle: jest.fn(),
    };

    const event = {
      preventDefault: jest.fn(),
    };

    const component = shallow(<EditArticle {...props} />);
    component.instance().saveOrPublish(event, { isDraft: false });
    component.instance().setState({
      ...article,
    });
    expect(component.state('title')).toEqual("Scientist don't die");
  });
});

describe('EditArticlePage component', () => {
  it('should match snapshot', () => {
    const wrapper = shallow(<EditArticleForm />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render without crashing', () => {
    const wrapper = shallow(<EditArticleForm />);
    expect(wrapper.find('div'));
    expect(wrapper.find('main'));
    expect(wrapper.find('form'));
    expect(wrapper.hasClass('title'));
    expect(wrapper.find('onClick'));
  });
});
