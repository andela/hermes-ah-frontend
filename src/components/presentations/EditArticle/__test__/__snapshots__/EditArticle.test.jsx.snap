// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditArticlePage component should match snapshot 1`] = `ShallowWrapper {}`;
